### PR TITLE
Add varchar and varbinary replicate function

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/VarbinaryFunctions.java [hunk 1]
```diff
@@ -472,4 +475,31 @@ public static Slice reverse(@SqlType("varbinary") Slice inputSlice)
         }
         return reverse;
     }
+
+    @Description("Repeats a given varbinary a number of times")
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/VarbinaryFunctions.java:479]:
```suggestion
    @Description("Repeats a varbinary given number of times")
```

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 2]
```diff
@@ -954,4 +965,24 @@ public static Slice translate(@SqlType("varchar(x)") Slice source, @SqlType("var
 
         return target;
     }
+
+    private static Slice repeat(Slice text, long times)
+    {
+        checkCondition(times >= 0, INVALID_FUNCTION_ARGUMENT, "Invalid number of string repetitions: " + times);
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java:971]: Is 0 valid?

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 3]
```diff
@@ -954,4 +965,24 @@ public static Slice translate(@SqlType("varchar(x)") Slice source, @SqlType("var
 
         return target;
     }
+
+    private static Slice repeat(Slice text, long times)
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java:969]: Inline?

File: presto-main/src/main/java/io/prestosql/operator/scalar/VarbinaryFunctions.java [hunk 4]
```diff
@@ -472,4 +475,31 @@ public static Slice reverse(@SqlType("varbinary") Slice inputSlice)
...
+    @ScalarFunction("replicate")
+    @SqlType(StandardTypes.VARBINARY)
+    public static Slice replicate(@SqlType("varbinary") Slice inputSlice, @SqlType(StandardTypes.BIGINT) long times)
+    {
+        if (inputSlice.length() == 0) {
+            return EMPTY_SLICE;
+        }
+
+        checkCondition(times >= 0, INVALID_FUNCTION_ARGUMENT, "Invalid number of string repetitions: " + times);
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/VarbinaryFunctions.java:488]: is 0 valid?

**Review by losipiuk — COMMENTED** [review]
@martint please take a look if syntax looks good to you

**Comment by losipiuk** [comment]
Also please add documentation as we settle that we want this function.

File: presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java [hunk 5]
```diff
@@ -726,6 +728,31 @@ public static Slice rightPad(@SqlType("varchar(x)") Slice text, @SqlType(Standar
...
+        try {
+            Slice buffer = Slices.allocate(size);
+            for (int i = 0; i < times; i++) {
+                buffer.setBytes(i * length, inputSlice, 0, length);
+            }
+
+            return buffer;
+        }
+        catch (SliceTooLargeException e) {
+            throw new PrestoException(EXCEEDED_FUNCTION_MEMORY_LIMIT, e.getMessage());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java:752]: We should cap the size at a more reasonable value. A 1GB allocation might work, for instance, but a page that big will cause all sorts of problems. I would pick a limit in the order of megabytes to align with the target size for a page. See PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES

**Review by martint — CHANGES_REQUESTED** [review]
I'd call the function `repeat`, but I'd like to hear other's thoughts.

**Comment by wendigo** [comment]
> I'd call the function `repeat`, but I'd like to hear other's thoughts.

@martint we already have repeat function for array creation with same signature

**Comment by martint** [comment]
> @martint we already have repeat function for array creation with same signature

Ugh, that's unfortunate, especially since other databases like MySQL, PostgreSQL and Oracle have a `repeat` function that works exactly like the one proposed by this change. The existing function should've never been added in that form.

We may want to deprecate the existing function and give it a different name, and introduce a new `repeat` that produces a varchar instead (similar to when we had to replace the `log` function after it was specified in SQL 2016). This will require waiting some time, as we need to give users enough time to adjust. It could also be gated behind a config option.

cc @prestosql/maintainers

**Comment by findepi** [comment]
@martint i agree.
Ideally, during the transition period we should have _no_ `repeat` function (at least by default).
The roadmap could look like

1. get agreement on the new name for `repeat(element, count) → array`  function
1. introduce new function as a replacement for `repeat(element, count) → array` and deprecate the existing one
    - then  users caring about warnings will have smooth upgrade path
1. (after some wait time) put deprecated `repeat(element, count) → array`  behind a toggle
    - then users not reading/getting warnings will have more rocky, but still manageable upgrade path 
    - when toggle is off, we can still expose `repeat(element, count) → array`, with some actionable err msg
1. (after some wait time) remove deprecated `repeat(element, count) → array`
1. (perhaps with no further delay) introduce new `repeat(varchar, count) → varchar` function

**Pull request closed without merging**